src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1234,25 +1380,30 @@ public static TSource Last<TSource>(this IEnumerable<TSource> source)
 
         public static TSource Last<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            if (predicate == null) throw Error.ArgumentNull("predicate");
-            TSource result = default(TSource);
-            bool found = false;
-            foreach (TSource element in source)
+            Error.CheckSourceNotNull(source);
+            Error.CheckPredicateNotNull(predicate);
+            using (IEnumerator<TSource> e = source.GetEnumerator())
             {
-                if (predicate(element))
+                while (e.MoveNext())
                 {
-                    result = element;
-                    found = true;
+                    TSource result = e.Current;
+                    if (predicate(result))
+                    {
+                        while (e.MoveNext())
+                        {
+                            TSource element = e.Current;
+                            if (predicate(element)) result = element;
+                        }
```

Comment:
Is this really better?  Yes, it saves writing one Boolean flag for each element that matches.  But it does so by duplicating the bulk of the function.  Is the difference really measurable?  I'd expect such a single Boolean write to a local to be dwarfed by the multiple interface calls and the delegate call each iteration, plus any time there's a write to the Boolean there's still a write to the local result.  Unless there's a real measurable difference, I'd prefer to stick with the simpler, more understandable foreach loop.